### ISSUE TITLE
Path handling toolkit

### ISSUE THREAD

**Issue by jennybc** [tidyverse/googledrive#24]
I have a second generation solution for resolving a requested path, such as `foo/bar/baz`.

I'd like to look ahead a bit and package it more generally, i.e. for use beyond `drive_ls()`.

@LucyMcGowan Will you help me list the ways other functions will need to work with paths?

`drive_list()`: User can supply path. We need to make sure such a path exists and is uniquely defined. Then determine file id for the terminal leaf and list it.

Functions that create or move files: We need to determine the maximal existing stem, in case some folders need to be created. Example: user asks to write a file to `foo/bar/baz/yo/myfile.txt`. We determine that `foo/bar` exists and that we need to create `baz` within `foo/bar` and `yo` within `foo/bar/baz`.

What else do we need?

**Comment by LucyMcGowan** [comment]
Those are the only two we currently use it for - off the top of my head I can't think of other use cases for needed to find a folder other than determining whether a path exists, if not, can we make it, it so, is it unique.

**Comment by jennybc** [comment]
OK great. Let us think about interface. A path goes in. What comes out? What do we need to do with the result?

It would be nice to have several small functions, but I think the API call burden would be too high. So I think we need one primary function that returns a list. What do we need to put into this list?

  * `n_paths` = How many such paths exist?
    - 0: path does not exist
    - 1: path exists and is uniquely defined
    - 2: more than one path exists, i.e. there are 2 or more `foo/bar/baz` paths that differ by at least one file id
  * `leaf_unique` is `TRUE` if path exists (possibly more than 1 such path!) and uniquely identifies the terminal leaf, `FALSE` otherwise
  * a list with one component per path or a tibble with one row per path?
  * a ?tibble? with one row per element of the path (split up) and info on (cumulative?) existence

Things get increasingly foggy as I go on. We need to figure what form of reporting will put as much logic as possible in this helper and list, so that our calls elsewhere are short and readable. Once I can take a first pass, we'll try it out and probably need to iterate a couple time.

**Comment by LucyMcGowan** [comment]
Would the list (or tibble) with one component per path contain the leafmost id?

**Comment by jennybc** [comment]
The leafmost id appears in the last two of the proposed objects up there. But not in a terribly accessible form. So yes that should be positioned for easier access.

In general, there's no guarantee there is exactly one leafmost id, right? I don't think this helper should throw many errors. I think the callers (or wrappers) have to use the return value to figure out whether to error.

Which reminds me: this will need to operate on paths that end in a folder and paths that end in a file, yes? Do you have any thoughts about that? The expectation probably needs to be part of the input and that affects the internal logic too, i.e. whether the first GET looks only for folders.

**Comment by jennybc** [comment]
After #34, all that remains to do here is to modify/extend the helper to better support functions like `drive_upload()`. Scenario: user says to upload to `foo/bar/baz/yo` but only `foo/bar` exists. We should be able to cope with that and create the folder `baz` in `foo/bar`, so we can make it the parent of the new file `yo`.

**Comment by LucyMcGowan** [comment]
I think there may be a small bug in the `get_leaf` function - when I run:
```
drive_mkdir("baz", path = "test_8675309/foo/yo")
```
I see
```
## Error: The path 'test_8675309/foo/yo/' identifies more than one file:
## File of type folder, with id 0B0Gh-SuuA2nTOEVQX1BteW1Fa3M.
## File of type folder, with id 0B0Gh-SuuA2nTN3VMYWdOZlhrM0k.
```
but on drive I have:
test_8675309
L foo
  L yo (id 0B0Gh-SuuA2nTOEVQX1BteW1Fa3M)
- yo (id 0B0Gh-SuuA2nTN3VMYWdOZlhrM0k)

so in fact test_8675309/foo/yo should be unique (since the other one that it's fussing about is actually test_8675309/yo)
